File: src/de/rub/bi/inf/openbimrl/functions/ifc/GetBoundingBox.java
package de.rub.bi.inf.openbimrl.functions.ifc;

import java.util.ArrayList;
import java.util.Collection;

import de.rub.bi.inf.openbimrl.NodeProxy;
import de.rub.bi.inf.openbimrl.engine.ifc.IIFCModel;
import de.rub.bi.inf.openbimrl.functions.AbstractFunction;

import javax.media.j3d.Bounds;
import javax.vecmath.Point3d;

/**
 * Retrieves the BoundingBox of an IFC element.
 * 
 * @author Marcel Stepien
 *
 */
public class GetBoundingBox extends AbstractFunction {

	public GetBoundingBox(NodeProxy nodeProxy) {
		super(nodeProxy);
	}

	@Override
	public void execute(IIFCModel ifcModel) {
	
		Object input0 = getInput(0);
		
		if(input0 == null)
			return;
				
		Collection<?> objects = null;
		if(input0 instanceof Collection<?>) {
			objects = (Collection<?>) input0;
		}else {
			ArrayList<Object> newList = new ArrayList<Object>();
			newList.add(input0);
			objects = newList;
		}
		
		ArrayList<Bounds> resultValues = new ArrayList<Bounds>();
		ArrayList<Point3d> centerValues = new ArrayList<Point3d>();
		
		for(Object o : objects) {
			ClassInterface classInterface = (ClassInterface) o;
			CadObject cad = ifcModel.getCadObjectModel().getCadObject(classInterface);
			
			Point3d center = new Point3d();
			Bounds b = cad.getBoundingBox();
			b.getCenter(center);
			
			resultValues.add(b);
			centerValues.add(center);
		}
		
		if(resultValues.size()==1) {
			setResult(0, resultValues.get(0));
		} else {
			setResult(0, resultValues);
		}

		if(centerValues.size()==1) {
			setResult(1, centerValues.get(0));
		} else {
			setResult(1, centerValues);
		}
	}

}

File: src/main/jvm/de/rub/bi/inf/openbimrl/functions/ifc/GetBoundingBox.kt
package de.rub.bi.inf.openbimrl.functions.ifc

import de.rub.bi.inf.nativelib.IfcPointer
import de.rub.bi.inf.nativelib.NativeEngine
import de.rub.bi.inf.openbimrl.NodeProxy
import de.rub.bi.inf.openbimrl.functions.NativeFunction
import de.rub.bi.inf.openbimrl.functions.annotations.FunctionPort
import de.rub.bi.inf.openbimrl.functions.annotations.OpenBIMRLFunction
import de.rub.bi.inf.openbimrl.utils.boundingBoxFromDoubles
import javax.media.j3d.BoundingBox
import javax.vecmath.Point3d

@OpenBIMRLFunction(
    description = "Retrieves the BoundingBox of an IFC element.",
    inputs = [
        FunctionPort(0, "IfcElement", IfcPointer::class),
    ],
    outputs = [
        FunctionPort(0, "Bounds", BoundingBox::class),
        FunctionPort(1, "CenterPoints", Point3d::class),
    ],
)
class GetBoundingBox(nodeProxy: NodeProxy) : NativeFunction(nodeProxy) {
    override fun executeNative() {
        when (val input = getInput<Any>(0)) {
            is IfcPointer -> applyBoundingBox(input)
            is Collection<*> -> {
                val elements = input.filterIsInstance<IfcPointer>()
                val centers = ArrayList<Point3d>(elements.size)
                val boxes = ArrayList<BoundingBox>(elements.size)
                elements.forEach { element ->
                    val bounds = NativeEngine.getBoundingBox(element.handle) ?: return@forEach
                    val (center, box) = boundingBoxFromDoubles(bounds)
                    centers.add(center)
                    boxes.add(box)
                }
                setResult(0, boxes)
                setResult(1, centers)
            }
        }
    }

    private fun applyBoundingBox(element: IfcPointer) {
        val bounds = NativeEngine.getBoundingBox(element.handle) ?: return
        val (center, box) = boundingBoxFromDoubles(bounds)
        setResult(0, box)
        setResult(1, center)
    }
}
